### PR TITLE
fix memleak, detach instead of clone to not drag around graph

### PR DESCRIPTION
**Pull request by janeyx99**
Thanks @clee2000 for bringing the memleak to my attention: https://github.com/pytorch/pytorch/actions/runs/12549765082/job/34996244798.

This memleak in the test was caused by the differentiable flavors. Because we had param.clone() and param persisted outside the for loop, the autograd graph would continue growing for each optimizer.step instead of being deleted after the optim input was used up.

To clarify, I had still expected (and still do expect) the test to fully clean everything up once the test is over, but I didn't get the chance to look into why that's not the case. This change would preliminarily unblock this particular test from failing the memleak CI.

Use detach instead of clone, which is...cheaper anyway :D since a detach I've learned from @soulitzer is a view with requires_grad=False

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #144154

### CODE REVIEW

**Review by clee2000 — APPROVED** [review]

**Review by Skylion007 — APPROVED** [review]

**Review by huydhn — APPROVED** [review]

**Comment by janeyx99** [comment]
@huydhn I am not convinced I've truly gotten to the bottom of this, because locally none of them fail but Adam and Adamax (not AdamW, Adagrad nor SGD) give warnings: 
```
/data/users/janeyx/pytorch/torch/testing/_internal/common_utils.py:2486: UserWarning: CUDA caching allocator reports a memory leak not verified by the driver API in __main__.TestOptimRenewedCUDA.test_grads_are_never_inplaced_into_Adam_cuda_float32! Caching allocator allocated memory was 0 and is now reported as 11776 on device 0. CUDA driver allocated memory was 1054539776 and is now 1052442624.
```

I notice that the leftover memory correspond to when we have to clone in the optimizer and when those intermediates are computationally related to the params. That was why I suspect it's the autograd graph keeping those tensors alive throughout the test. As I also updated above in my PR body, I believe I fixed this test from displaying memory leakiness, but I am not sure why the entire test being concluded did not automatically remove that memory.

**Comment by vadimkantorov** [comment]
I hope the `.clone()` of step/betas will not lead to the same (maybe clone of beta could?)

**Review by albanD — APPROVED** [review]
Ok!
